src/Microsoft.Performance.SDK.Runtime.NetCoreApp/Plugins/PluginsLoader.cs

```diff
@@ -0,0 +1,255 @@
+        ///     sources being loaded has already been loaded by either the current or a previous call to <see cref="TryLoadPlugin(string)"/>
+        ///     or <see cref="TryLoadPlugins(IEnumerable{string}, out IList{string})"/>.
+        /// </remarks>
+        public bool TryLoadPlugins(IEnumerable<string> directories, out IList<string> failed)
+        {
+            lock (this.mutex)
+            {
+                failed = new List<string>();
+                var oldPlugins = new HashSet<CustomDataSourceReference>(this.catalog.PlugIns);
+                foreach (var dir in directories)
+                {
+
+                    if (!this.extensionDiscovery.ProcessAssemblies(dir))
+                    {
+                        failed.Add(dir);
+                        continue;
+                    }
+                }
+
+                // TODO: this does redundant work re-finalizing custom data sources
+                // loaded by previous calls to this method. The extension repository
+                // should get refactored to avoid this redundant work.
+                this.extensionRepository.FinalizeDataExtensions();
+
+                foreach (var source in this.catalog.PlugIns.Except(oldPlugins))
+                {
+                    var info = this.GetPluginNameAndVersion(source);
+                    this.NotifyCustomDataSourceLoaded(info.Item1, info.Item2, source);
+                }
+
+                return failed.Count == 0;
+            }
+        }
+
+        /// <summary>
+        ///     Registers an <see cref="IPluginsConsumer"/> to receive all future plugins, and sends all
+        ///     previously loaded plugins to its <see cref="IPluginsConsumer.OnCustomDataSourceLoaded(string, CustomDataSourceReference)"/> handler.
+        ///     <para/>
+        ///     While the consumer is receiving all previously loaded plugins, this method blocks and
+        ///     plugins cannot be loaded.
+        /// </summary>
+        /// <param name="consumer">The consumer to subscribe and receive all loaded plugins</param>
+        public void Subscribe(IPluginsConsumer consumer)
+        {
+            lock (this.mutex)
+            {
+                // Manually send all the already loaded plugins to this consumer
+                foreach (var source in this.catalog.PlugIns)
+                {
+                    var info = this.GetPluginNameAndVersion(source);
+                    consumer.OnCustomDataSourceLoaded(info.Item1, info.Item2, source);
+                }
+
+                // Subscribe to all future plugin loads
+                this.subscribers.Add(consumer);
+            }
+        }
+
+        /// <summary>
+        ///     Unregisters an <see cref="IPluginsConsumer"/> from plugins loaded in the future.
+        /// </summary>
+        /// <param name="consumer"></param>
+        public void Unsubscribe(IPluginsConsumer consumer)
+        {
+            this.subscribers.Remove(consumer);
```

Comment:
>Remove [](start = 29, length = 6)

since we already have a mutex, I'm wondering if we should stop using ConcurrentSet and *only* use the lock. I know this is a pretty extreme example, but imagine a multithreaded caller that calls Subscribe and Unsubscribe on multiple threads. subscribe starts enumerating existing plug-ins, unsubscribe executes on a different thread, then the subscriber is added. I realize that using only the lock doesn't solve all of cases with this, as unsubscribe could run *first*, but in the case described before, if we're only using a lock, then unsubscribe would wait for subscribe to finish before executing.

also, might be worth returning a bool from unsubscribe that indicates whether the subscriber was removed. if nothing else, this might provide an indication that something didn't work as expected for the develop.

i'm not saying that we must make this change, but i believe it's worth considering.

Reply:
In the case you listed above, the observed behavior would be as if subscribe happened entirely before unsubscribe, which is a valid ordering. The class only guaranteed calls are executed in a linearizable order, and the unsubscribe only guaranteed they do not observe _future_ loaded plugins. Plugins loaded concurrently (i.e. before Unsubscribe returns) or before the call are fine for them to see. 

I agree that we should return a bool indicating if they were successfully unsubscribed, since I can see that being an edge case if the caller is subscribing concurrently (or if they never subscribed at all).

Reply:
In the case you listed above, the observed behavior would be as if subscribe happened entirely before unsubscribe, which is a valid ordering. The class only guaranteed calls are executed in a linearizable order, and the unsubscribe only guaranteed they do not observe _future_ loaded plugins. Plugins loaded concurrently (i.e. before Unsubscribe returns) or before the call are fine for them to see. 

I agree that we should return a bool indicating if they were successfully unsubscribed, since I can see that being an edge case if the caller is subscribing concurrently (or if they never subscribed at all). 

EDIT: I missed that last part about the actual subscribe being added after unsubscribe returns. In that case it would actually appear as if unsubscribe finished entirely before subscribe, even though the consumer observes the plugins before the collection is modified. The bool would mitigate this issue, but I see the value in using a lock here to prevent that confusion.